<h1>What is Pandas?</h1>

Pandas is a Python library used for working with data sets.

It has functions for analyzing, cleaning, exploring, and manipulating data.

<b>Why Use Pandas?</b>

Pandas allows us to analyze big data and make conclusions based on statistical theories.

Pandas can clean messy data sets, and make them readable and relevant.

Relevant data is very important in data science.


<b>What Can Pandas Do?</b>

Pandas gives you answers about the data. Like:

Is there a correlation between two or more columns?

What is average value?

Max value?

Min value?

Pandas are also able to delete rows that are not relevant, or contains wrong values, like empty or NULL values. This is called cleaning the data.

Import Pandas

Once Pandas is installed, import it in your applications by adding the import keyword:

In [3]:
import pandas # pip install pandas if you are getting error

In [6]:
mydataset = {'cars' :['BMW','Volvo','Ford'],
             'passings':[3,7,2]}
myvar = pandas.DataFrame(mydataset)
myvar

cars  passings
0    BMW         3
1  Volvo         7
2   Ford         2

<b>Pandas as pd</b>

Pandas is usually imported under the pd alias.

alias: In Python alias are an alternate name for referring to the same thing.

Create an alias with the as keyword while importing:

In [8]:
import pandas as pd
mydataset = {'cars' :['BMW','Volvo','Ford'],
             'passings':[3,2,2]
    }
myvar = pd.DataFrame(mydataset)
myvar

cars  passings
0    BMW         3
1  Volvo         2
2   Ford         2

<h1>Pandas Series</h1>

What is a Series?

A Pandas Series is like a column in a table.

It is a one-dimensional array holding data of any type.

Example

Create a simple Pandas Series from a list:

In [12]:
a = [555,25,89,33,555]
myvar =pd.Series(a)
myvar

0    555
1     25
2     89
3     33
4    555
dtype: int64

<b>Labels</b>

If nothing else is specified, the values are labeled with their index number. First value has index 0, second value has index 1 etc.

This label can be used to access a specified value.

Example

Return the first value of the Series:

In [13]:
print(myvar[4])

555


<b>Create Labels</b>

With the index argument, you can name your own labels.

Example

Create you own labels:

In [15]:
a =[1,222,568]
myvar = pd.Series(a,index =['x','y','z'])
myvar

x      1
y    222
z    568
dtype: int64

In [16]:
print(myvar['z'])

568


<b>Key/Value Objects as Series</b>

You can also use a key/value object, like a dictionary, when creating a Series.

Example

Create a simple Pandas Series from a dictionary:

The keys of the dictionary become the labels.

In [22]:
cal = {'day1':43,'day2':350,'day3':390}
myvar = pd.Series(cal)
print(myvar)

day1     43
day2    350
day3    390
dtype: int64


To select only some of the items in the dictionary, use the index argument and specify only the items you want to include in the Series.

Example

Create a Series using only data from "day1" and "day2":





In [23]:
cal = {'day1':43,'day2':350,'day3':390}
myvar = pd.Series(cal,index =['day1','day2'])
print(myvar)

day1     43
day2    350
dtype: int64


<B>What is a DataFrame?</B>

A Pandas DataFrame is a 2 dimensional data structure, like a 2 dimensional array, or a table with rows and columns.

Example
Create a simple Pandas DataFrame:

In [24]:
data = {'calories':[420,380,390],
       'duration':[50,40,80]}
df =pd.DataFrame(data)
print(df)

   calories  duration
0       420        50
1       380        40
2       390        80


<B>Locate Row</B>

As you can see from the result above, the DataFrame is like a table with rows and columns.

Pandas use the loc attribute to return one or more specified row(s)

Example

Return row 0:

In [34]:
print(df.loc[[2,0]])

   calories  duration
2       390        80
0       420        50


<b>Named Indexes</b>

With the index argument, you can name your own indexes.

Example
Add a list of names to give each row a name:

In [35]:
data = {'calories':[420,380,390],
       'duration':[50,40,80]}
df =pd.DataFrame(data,index = ['day1','day2','day3'])
print(df)

      calories  duration
day1       420        50
day2       380        40
day3       390        80


<b>Locate Named Indexes</b>

Use the named index in the loc attribute to return the specified row(s).

Example

Return "day2":

In [36]:
print(df.loc['day2'])


calories    380
duration     40
Name: day2, dtype: int64


<b>Load Files Into a DataFrame</b>

If your data sets are stored in a file, Pandas can load them into a DataFrame.

Example

Load a comma separated file (CSV file) into a DataFrame:

In [9]:
df = pd.read_csv('C:\Seema Python CSV\data.csv')
print(df)

    linear  squared    cubed  power_of_two
0        1        1        1  2.000000e+00
1        2        4        8  4.000000e+00
2        3        9       27  8.000000e+00
3        4       16       64  1.600000e+01
4        5       25      125  3.200000e+01
..     ...      ...      ...           ...
95      96     9216   884736  7.922820e+28
96      97     9409   912673  1.584560e+29
97      98     9604   941192  3.169130e+29
98      99     9801   970299  6.338250e+29
99     100    10000  1000000  1.267650e+30

[100 rows x 4 columns]


<b>Read CSV Files</b>

A simple way to store big data sets is to use CSV files (comma separated files).

CSV files contains plain text and is a well know format that can be read by everyone including Pandas.

In our examples we will be using a CSV file called 'data.csv

Example:
Load the CSV into a DataFrame:

*use to_string() to print the entire DataFrame.

In [11]:
df = pd.read_csv('C:\Seema Python CSV\data.csv')
print(df.to_string())

    linear  squared    cubed  power_of_two
0        1        1        1  2.000000e+00
1        2        4        8  4.000000e+00
2        3        9       27  8.000000e+00
3        4       16       64  1.600000e+01
4        5       25      125  3.200000e+01
5        6       36      216  6.400000e+01
6        7       49      343  1.280000e+02
7        8       64      512  2.560000e+02
8        9       81      729  5.120000e+02
9       10      100     1000  1.024000e+03
10      11      121     1331  2.048000e+03
11      12      144     1728  4.096000e+03
12      13      169     2197  8.192000e+03
13      14      196     2744  1.638400e+04
14      15      225     3375  3.276800e+04
15      16      256     4096  6.553600e+04
16      17      289     4913  1.310720e+05
17      18      324     5832  2.621440e+05
18      19      361     6859  5.242880e+05
19      20      400     8000  1.048576e+06
20      21      441     9261  2.097152e+06
21      22      484    10648  4.194304e+06
22      23 

<b>Pandas - Analyzing DataFrames</b>

Viewing the Data

One of the most used method for getting a quick overview of the DataFrame, is the head() method.

The head() method returns the headers and a specified number of rows, starting from the top.

Example
Get a quick overview by printing the first 10 rows of the DataFrame:

In [42]:
df.head(9)

linear  squared  cubed  power_of_two
0       1        1      1           2.0
1       2        4      8           4.0
2       3        9     27           8.0
3       4       16     64          16.0
4       5       25    125          32.0
5       6       36    216          64.0
6       7       49    343         128.0
7       8       64    512         256.0
8       9       81    729         512.0

There is also a tail() method for viewing the last rows of the DataFrame.

The tail() method returns the headers and a specified number of rows, starting from the bottom.

Example

Print the last 5 rows of the DataFrame:

In [44]:
df.tail(10)

linear  squared    cubed  power_of_two
90      91     8281   753571  2.475880e+27
91      92     8464   778688  4.951760e+27
92      93     8649   804357  9.903520e+27
93      94     8836   830584  1.980700e+28
94      95     9025   857375  3.961410e+28
95      96     9216   884736  7.922820e+28
96      97     9409   912673  1.584560e+29
97      98     9604   941192  3.169130e+29
98      99     9801   970299  6.338250e+29
99     100    10000  1000000  1.267650e+30

<b>Info About the Data</b>

The DataFrames object has a method called info(), that gives you more information about the data set.

Example:
Print information about the data:

In [45]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   linear        100 non-null    int64  
 1   squared       100 non-null    int64  
 2   cubed         100 non-null    int64  
 3   power_of_two  100 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 3.2 KB


<b>Null Values</b>

The info() method also tells us how many Non-Null values there are present in each column, and in our data set it seems like there are 164 of 169 Non-Null values in the "Calories" column.

Which means that there are 5 rows with no value at all, in the "Calories" column, for whatever reason.

Empty values, or Null values, can be bad when analyzing data, and you should consider removing rows with empty values. This is a step towards what is called cleaning data.

<b>Data Cleaning</b>

Data cleaning means fixing bad data in your data set.

Bad data could be:

Empty cells

Data in wrong format

Wrong data

Duplicates

Empty cells can potentially give you a wrong result when you analyze data.

Remove Rows :
One way to deal with empty cells is to remove rows that contain empty cells.

This is usually OK, since data sets can be very big, and removing a few rows will not have a big impact on the result.

Example:
Return a new Data Frame with no empty cells:

In [12]:
df = pd.read_csv('C:\Seema Python CSV\dirtydata.csv')
new_df = df.dropna()
print(new_df)

    Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'  

Note: By default, the dropna() method returns a new DataFrame, and will not change the original.

If you want to change the original DataFrame, use the inplace = True argument:

Example: 
Remove all rows with NULL values:

In [56]:
df = pd.read_csv('C:\Seema Python CSV\dirtydata.csv')
df.dropna(inplace = True)
print(df)

    Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'  

<b>Replace Empty Values</b>

Another way of dealing with empty cells is to insert a new value instead.

This way you do not have to delete entire rows just because of some empty cells.

The fillna() method allows us to replace empty cells with a value:

Example:
Replace NULL values with the number 130:

In [59]:
df = pd.read_csv('C:\Seema Python CSV\dirtydata.csv')
df.fillna(150,inplace =True)
df

Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'    100       120     300.0
18        45  '2020/12/18'     90       112     150.0
19        60  '2020/12/19'    103       123     323.0
20        45  '2020/12/20'     97       125     243.0
21        60  '2020/12/21'    108       131     364.2
22        45           150    100       119     282.0
23        60  '2020/12/23'    130       101     300.0
24        45  '2020/12/24'    105       132     246.0
25        60  '2020/12/25'    102       126     334.5
26        60      20201226    100       120     250.0
27        60  '2020/12/27'     92       118     241.0
28        60  '2020/12/28'    103       132     150.0
29        60  '2020/12/29'    100       132     280.0
30        60  '2020/12/30'    102       129     380.3
31        60  '2020/12/31'     92       115     243.0

<b>Replace Only For a Specified Columns</b>

The example above replaces all empty cells in the whole Data Frame.

To only replace empty values for one column, specify the column name for the DataFrame:

Example:
Replace NULL values in the "Calories" columns with the number 130:

In [60]:
df = pd.read_csv('C:\Seema Python CSV\dirtydata.csv')
df['Calories'].fillna(130,inplace=True)
print(df)

    Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'  

<b>Replace Using Mean, Median, or Mode</b>

A common way to replace empty cells, is to calculate the mean, median or mode value of the column.

Pandas uses the mean() median() and mode() methods to calculate the respective values for a specified column:

Example:Calculate the MEAN, and replace any empty values with it:

In [73]:
df = pd.read_csv('C:\Seema Python CSV\dirtydata.csv')

x =df['Calories'].mean()
print(x)
df['Calories'].fillna(x,inplace = True)
df

304.68


Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130    409.10
1         60  '2020/12/02'    117       145    479.00
2         60  '2020/12/03'    103       135    340.00
3         45  '2020/12/04'    109       175    282.40
4         45  '2020/12/05'    117       148    406.00
5         60  '2020/12/06'    102       127    300.00
6         60  '2020/12/07'    110       136    374.00
7        450  '2020/12/08'    104       134    253.30
8         30  '2020/12/09'    109       133    195.10
9         60  '2020/12/10'     98       124    269.00
10        60  '2020/12/11'    103       147    329.30
11        60  '2020/12/12'    100       120    250.70
12        60  '2020/12/12'    100       120    250.70
13        60  '2020/12/13'    106       128    345.30
14        60  '2020/12/14'    104       132    379.30
15        60  '2020/12/15'     98       123    275.00
16        60  '2020/12/16'     98       120    215.20
17        60  '2020/12/17'    100       120    300.00
18        45  '2020/12/18'     90       112    304.68
19        60  '2020/12/19'    103       123    323.00
20        45  '2020/12/20'     97       125    243.00
21        60  '2020/12/21'    108       131    364.20
22        45           NaN    100       119    282.00
23        60  '2020/12/23'    130       101    300.00
24        45  '2020/12/24'    105       132    246.00
25        60  '2020/12/25'    102       126    334.50
26        60      20201226    100       120    250.00
27        60  '2020/12/27'     92       118    241.00
28        60  '2020/12/28'    103       132    304.68
29        60  '2020/12/29'    100       132    280.00
30        60  '2020/12/30'    102       129    380.30
31        60  '2020/12/31'     92       115    243.00

Mean = the average value (the sum of all values divided by number of values).

Example:
Calculate the MEDIAN, and replace any empty values with it:

In [102]:
df = pd.read_csv('C:\Seema Python CSV\dirtydata.csv')

x =df['Calories'].median()
print(x)
df['Calories'].fillna(x,inplace = True)
df

291.2


Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'    100       120     300.0
18        45  '2020/12/18'     90       112     291.2
19        60  '2020/12/19'    103       123     323.0
20        45  '2020/12/20'     97       125     243.0
21        60  '2020/12/21'    108       131     364.2
22        45           NaN    100       119     282.0
23        60  '2020/12/23'    130       101     300.0
24        45  '2020/12/24'    105       132     246.0
25        60  '2020/12/25'    102       126     334.5
26        60      20201226    100       120     250.0
27        60  '2020/12/27'     92       118     241.0
28        60  '2020/12/28'    103       132     291.2
29        60  '2020/12/29'    100       132     280.0
30        60  '2020/12/30'    102       129     380.3
31        60  '2020/12/31'     92       115     243.0

Median = the value in the middle, after you have sorted all values ascending.

Example
Calculate the MODE, and replace any empty values with it:

In [2]:
import pandas as pd
df = pd.read_csv('C:\Seema Python CSV\dirtydata.csv')

x =df['Calories'].mode()[0]
print(x)
df['Calories'].fillna(x,inplace = True)
df

300.0


Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'    100       120     300.0
18        45  '2020/12/18'     90       112     300.0
19        60  '2020/12/19'    103       123     323.0
20        45  '2020/12/20'     97       125     243.0
21        60  '2020/12/21'    108       131     364.2
22        45           NaN    100       119     282.0
23        60  '2020/12/23'    130       101     300.0
24        45  '2020/12/24'    105       132     246.0
25        60  '2020/12/25'    102       126     334.5
26        60      20201226    100       120     250.0
27        60  '2020/12/27'     92       118     241.0
28        60  '2020/12/28'    103       132     300.0
29        60  '2020/12/29'    100       132     280.0
30        60  '2020/12/30'    102       129     380.3
31        60  '2020/12/31'     92       115     243.0

<h1>Pandas - Cleaning Data of Wrong Format</h1>

Data of Wrong Format:
    
Cells with data of wrong format can make it difficult, or even impossible, to analyze data.

To fix it, you have two options: remove the rows, or convert all cells in the columns into the same format.

Convert Into a Correct Format
In our Data Frame, we have two cells with the wrong format. Check out row 22 and 26, the 'Date' column should be a string that represents a date:

Let's try to convert all cells in the 'Date' column into dates.

Pandas has a to_datetime() method for this:

Example
Convert to date:



In [13]:
import pandas as pd
df = pd.read_csv("C:\Seema Python CSV\dirtydata.csv")
df['Date'] = pd.to_datetime(df['Date'])
print(df.to_string())

    Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7        450 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90 

As you can see from the result, the date in row 26 where fixed, but the empty date in row 22 got a NaT (Not a Time) value, in other words an empty value. One way to deal with empty values is simply removing the entire row.

Removing Rows

The result from the converting in the example above gave us a NaT value, which can be handled as a NULL value, and we can remove the row by using the dropna() method.

Example
Remove rows with a NULL value in the "Date" column:

In [15]:
df['Date'].dropna(inplace = True)
df

Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7        450 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90       112       NaN
19        60 2020-12-19    103       123     323.0
20        45 2020-12-20     97       125     243.0
21        60 2020-12-21    108       131     364.2
22        45        NaT    100       119     282.0
23        60 2020-12-23    130       101     300.0
24        45 2020-12-24    105       132     246.0
25        60 2020-12-25    102       126     334.5
26        60 2020-12-26    100       120     250.0
27        60 2020-12-27     92       118     241.0
28        60 2020-12-28    103       132       NaN
29        60 2020-12-29    100       132     280.0
30        60 2020-12-30    102       129     380.3
31        60 2020-12-31     92       115     243.0

<b>Wrong Data</b>

"Wrong data" does not have to be "empty cells" or "wrong format", it can just be wrong, like if someone registered "199" instead of "1.99".

Sometimes you can spot wrong data by looking at the data set, because you have an expectation of what it should be.

If you take a look at our data set, you can see that in row 7, the duration is 450, but for all the other rows the duration is between 30 and 60.

It doesn't have to be wrong, but taking in consideration that this is the data set of someone's workout sessions, we conclude with the fact that this person did not work out in 450 minutes.


How can we fix wrong values, like the one for "Duration" in row 7?

<b>Replacing Values:</b>

One way to fix wrong values is to replace them with something else.

In our example, it is most likely a typo, and the value should be "45" instead of "450", and we could just insert "45" in row 7:

Example:
Set "Duration" = 45 in row 7:

In [18]:
df.loc[7,'Duration'] = 45
df

Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7         45 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90       112       NaN
19        60 2020-12-19    103       123     323.0
20        45 2020-12-20     97       125     243.0
21        60 2020-12-21    108       131     364.2
22        45        NaT    100       119     282.0
23        60 2020-12-23    130       101     300.0
24        45 2020-12-24    105       132     246.0
25        60 2020-12-25    102       126     334.5
26        60 2020-12-26    100       120     250.0
27        60 2020-12-27     92       118     241.0
28        60 2020-12-28    103       132       NaN
29        60 2020-12-29    100       132     280.0
30        60 2020-12-30    102       129     380.3
31        60 2020-12-31     92       115     243.0

For small data sets you might be able to replace the wrong data one by one, but not for big data sets.

To replace wrong data for larger data sets you can create some rules, e.g. set some boundaries for legal values, and replace any values that are outside of the boundaries.

Example:
Loop through all values in the "Duration" column.

If the value is higher than 120, set it to 120:

In [19]:
for x in df.index:
    if df.loc[x,'Duration']>120:
        df.loc[x,"Duration"] = 120

<b>Removing Rows</b>
Another way of handling wrong data is to remove the rows that contains wrong data.

This way you do not have to find out what to replace them with, and there is a good chance you do not need them to do your analyses.

Example
Delete rows where "Duration" is higher than 120:

In [20]:
for x in df.index:
    if df.loc[x,'Duration']>120:
        df.drop(x,inplace =True)

<b>Discovering Duplicates</b>

Duplicate rows are rows that have been registered more than one time.

By taking a look at our test data set, we can assume that row 11 and 12 are duplicates.

To discover duplicates, we can use the duplicated() method.

The duplicated() method returns a Boolean values for each row:

Example
Returns True for every row that is a duplicate, othwerwise False:

In [21]:
df

Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7         45 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90       112       NaN
19        60 2020-12-19    103       123     323.0
20        45 2020-12-20     97       125     243.0
21        60 2020-12-21    108       131     364.2
22        45        NaT    100       119     282.0
23        60 2020-12-23    130       101     300.0
24        45 2020-12-24    105       132     246.0
25        60 2020-12-25    102       126     334.5
26        60 2020-12-26    100       120     250.0
27        60 2020-12-27     92       118     241.0
28        60 2020-12-28    103       132       NaN
29        60 2020-12-29    100       132     280.0
30        60 2020-12-30    102       129     380.3
31        60 2020-12-31     92       115     243.0

In [22]:
print(df.duplicated())

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12     True
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
22    False
23    False
24    False
25    False
26    False
27    False
28    False
29    False
30    False
31    False
dtype: bool


Removing Duplicates:
To remove duplicates, use the drop_duplicates() method.

Example:
Remove all duplicates:

The (inplace = True) will make sure that the method does NOT return a new DataFrame, but it will remove all duplicates from the original DataFrame.

In [25]:
df.drop_duplicates(inplace = True)
df

Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7         45 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90       112       NaN
19        60 2020-12-19    103       123     323.0
20        45 2020-12-20     97       125     243.0
21        60 2020-12-21    108       131     364.2
22        45        NaT    100       119     282.0
23        60 2020-12-23    130       101     300.0
24        45 2020-12-24    105       132     246.0
25        60 2020-12-25    102       126     334.5
26        60 2020-12-26    100       120     250.0
27        60 2020-12-27     92       118     241.0
28        60 2020-12-28    103       132       NaN
29        60 2020-12-29    100       132     280.0
30        60 2020-12-30    102       129     380.3
31        60 2020-12-31     92       115     243.0

In [24]:
df['Duration'].unique()

array([60, 45, 30], dtype=int64)

<h2>Pandas - Data Correlations</h2>

Finding Relationships

A great aspect of the Pandas module is the corr() method.

The corr() method calculates the relationship between each column in your data set.

Example
Show the relationship between the columns:

In [26]:
df = pd.read_csv("C:\Seema Python CSV\dirtydata.csv")
df.corr()

Duration     Pulse  Maxpulse  Calories
Duration  1.000000  0.004410  0.049959 -0.114169
Pulse     0.004410  1.000000  0.276583  0.513186
Maxpulse  0.049959  0.276583  1.000000  0.357460
Calories -0.114169  0.513186  0.357460  1.000000


The Result of the corr() method is a table with a lot of numbers that represents how well the relationship is between two columns.

The number varies from -1 to 1.

1 means that there is a 1 to 1 relationship (a perfect correlation), and for this data set, each time a value went up in the first column, the other one went up as well.

0.9 is also a good relationship, and if you increase one value, the other will probably increase as well.

-0.9 would be just as good relationship as 0.9, but if you increase one value, the other will probably go down.

0.2 means NOT a good relationship, meaning that if one value goes up does not mean that the other will.

What is a good correlation? It depends on the use, but I think it is safe to say you have to have at least 0.6 (or -0.6) to call it a good correlation.

Perfect Correlation:
We can see that "Duration" and "Duration" got the number 1.000000, which makes sense, each column always has a perfect relationship with itself.

Good Correlation:
"Duration" and "Calories" got a 0.922721 correlation, which is a very good correlation, and we can predict that the longer you work out, the more calories you burn, and the other way around: if you burned a lot of calories, you probably had a long work out.

Bad Correlation:
"Duration" and "Maxpulse" got a 0.009403 correlation, which is a very bad correlation, meaning that we can not predict the max pulse by just looking at the duration of the work out, and vice versa.



DataFrame Basic Functionality
Let us now understand what DataFrame Basic Functionality is. The following tables lists down the important attributes or methods that help in DataFrame Basic Functionality.

T : Transposes rows and columns.

axes : Returns a list with the row axis labels and column axis labels as the only members.


dtypes : Returns the dtypes in this object.

empty : True if NDFrame is entirely empty [no items]; if any of the axes are of length 0.
	
ndim : Number of axes / array dimensions.

shape : Returns a tuple representing the dimensionality of the DataFrame.

size : Number of elements in the NDFrame.

values : Numpy representation of NDFrame.

head() : Returns the first n rows.

tail() : Returns last n rows.

Let us now create a DataFrame and see all how the above mentioned attributes operate.

Example:

In [27]:
import pandas as pd
import numpy as np
d = {"Name" : pd.Series(['Jack','Ricky','Vin','Steve','Smith','Tom','James']),
    'Age':pd.Series([25,26,27,28,30,33,65]),
    'Rating' : pd.Series([4.2,6.2,8.5,9.3,5.5,3.3,2.2])}

df = pd.DataFrame(d)
print(df)

    Name  Age  Rating
0   Jack   25     4.2
1  Ricky   26     6.2
2    Vin   27     8.5
3  Steve   28     9.3
4  Smith   30     5.5
5    Tom   33     3.3
6  James   65     2.2


<b>T (Transpose)</b>

Returns the transpose of the DataFrame. The rows and columns will interchange.

In [22]:
df.T

0      1    2      3      4    5      6
Name    Jack  Ricky  Vin  Steve  Smith  Tom  James
Age       25     26   27     28     30   33     65
Rating   4.2    6.2  8.5    9.3    5.5  3.3    2.2

<b>axes</b>

Returns the list of row axis labels and column axis labels.

In [28]:
df.axes

[RangeIndex(start=0, stop=7, step=1),
 Index(['Name', 'Age', 'Rating'], dtype='object')]

<b>dtypes</b>
Returns the data type of each column.

In [29]:
df.dtypes

Name       object
Age         int64
Rating    float64
dtype: object

<b>empty</b>
Returns the Boolean value saying whether the Object is empty or not; True indicates that the object is empty.

In [30]:
df.empty

False

<b>ndim</b>
Returns the number of dimensions of the object. By definition, DataFrame is a 2D object.

In [26]:
df.ndim

2

In [32]:
df

Name  Age  Rating
0   Jack   25     4.2
1  Ricky   26     6.2
2    Vin   27     8.5
3  Steve   28     9.3
4  Smith   30     5.5
5    Tom   33     3.3
6  James   65     2.2

<b>shape</b>
Returns a tuple representing the dimensionality of the DataFrame. Tuple (a,b), where a represents the number of rows and b represents the number of columns.

In [35]:
df.shape

(7, 3)

<b>size</b>
Returns the number of elements in the DataFrame.

In [31]:
df.size

21

<b>values</b>
Returns the actual data in the DataFrame as an NDarray.

In [36]:
df.values

array([['Jack', 25, 4.2],
       ['Ricky', 26, 6.2],
       ['Vin', 27, 8.5],
       ['Steve', 28, 9.3],
       ['Smith', 30, 5.5],
       ['Tom', 33, 3.3],
       ['James', 65, 2.2]], dtype=object)

<b>The describe() Function^</b>

It returns all the necessary values such as the count, min, max, mean, median etc. for a series.

In [37]:
df.describe()

Age    Rating
count   7.000000  7.000000
mean   33.428571  5.600000
std    14.175767  2.622975
min    25.000000  2.200000
25%    26.500000  3.750000
50%    28.000000  5.500000
75%    31.500000  7.350000
max    65.000000  9.300000

As you can see, we have got the summary of descriptive statistics for the `df` series. You can also spot some of the values against the `25%, 50%` and `75%` parameters. They are called percentiles (or quartiles). Let's learn each of them one-by-one. Let's start with `25%`.

**Percentiles (or Quartiles)**

`25%` refers to the 
$25^{th}$ 
percentile value. It is also called the **first quartile** value. *The first quartile is defined as the middle number between the smallest number and the median of the dataset.* 

Let's understand this concept through an example. Consider the following series of numbers,

$3, 9, 6, 10, 7, 9, 7, 6, 1$

After arranging them in the increasing order, we have the following series

$1, 3, 6, 6, 7, 7, 9, 9, 10$

The median value of the series is 
$7$ 
because it is the middle value. So, the 
$25^{th}$ 
percentile (or first quartile) value will be the median of the following values

$1, 3, 6, 6, 7$

Therefore, first quartile or 
$Q1 = 6$ 
because 
$6$ 
is the median of the above sub-series.

**Note:** The median value is also referred to as the **second quartile** (or 
$50^{th}$ 
percentile) value.

Similarly, `75%` refers to the 
$75^{th}$ 
percentile value. It is also called the **third quartile** value. *The third quartile is the middle value between the median and the highest value of the dataset.*

So, the 
$75^{th}$ 
percentile (or third quartile) value will be the median of the following values

$7, 7, 9, 9, 10$

Therefore, third quartile or 
$Q3 = 9$ 
because 
$9$ 
is the median of the above sub-series.

*The first, second and third quartile values together tell the spread of the data. If a majority of the data points are spread apart from each other, then there will be a significantly large difference between the three quartile values whereas if a majority of the data points are close to each other, then all three quartile values will also be close to each other.*

In the 

$3, 9, 6, 10, 7, 9, 7, 6, 1$

series, the first, second and third quartile values are 
$Q1=6, Q2=7$ 
and 
$Q3=9$ 
respectively. This means the majority of the data points are quite close to each other which is quite evident from the series.


**Standard Deviation (or `std`)**

*The standard deviation is the square root of the mean of squared deviations.* It tells us how far away a data point is from the mean value. Larger the standard deviation, farther the data point from the mean value.

For the above series, the mean value
$(\mu)$
is 

$\mu = \frac{3+9+6+10+7+9+7+6+1}{9} = 6.44$

The **deviation** is the difference between the mean value and the individual data points in a series, i.e., 

$(3 - 6.44), (9 - 6.44), (6 - 6.44), (10 - 6.44), (7 - 6.44), (9 - 6.44), (7 - 6.44), (6 - 6.44), (1 - 6.44)$

are the deviations in the series. 

The **squared deviation** is the square of deviations, i.e., 

$(3 - 6.44)^2, (9 - 6.44)^2, (6 - 6.44)^2, (10 - 6.44)^2, (7 - 6.44)^2, (9 - 6.44)^2, (7 - 6.44)^2, (6 - 6.44)^2, (1 - 6.44)^2$

are the squared deviations in the series.

The **mean of squared deviations (or MSD)** is a sum of all the squared deviations divided by the number number of values, i.e., 

$\frac{(3 - 6.44)^2 + (9 - 6.44)^2 + (6 - 6.44)^2 + (10 - 6.44)^2 + (7 - 6.44)^2 + (9 - 6.44)^2 + (7 - 6.44)^2 + (6 - 6.44)^2 + (1 - 6.44)^2}{9}$

So, the standard deviation (or 
$std$
) is

$\sqrt{\frac{(3 - 6.44)^2 + (9 - 6.44)^2 + (6 - 6.44)^2 + (10 - 6.44)^2 + (7 - 6.44)^2 + (9 - 6.44)^2 + (7 - 6.44)^2 + (6 - 6.44)^2 + (1 - 6.44)^2}{9}}$

$\Rightarrow std = 2.92$

The significance of the standard deviation is that the numbers, 
$9, 6, 7, 9, 7, 6$
, from the series, lie in the range 
$(mean - std)$
and
$(mean + std)$
, i.e., 
$6.44 - 2.92 = 3.52$
and
$6.44 + 2.92 = 9.36$

So, these numbers are closest to the mean value.

The remaining three numbers, 
$1, 3$
and
$10$
lie in the range
$(mean - 2\times std)$
and
$(mean + 2\times std)$
. So, these three numbers are farthest from the mean value.

We can verify the above calculations by first creating a Pandas series using the `pd.Series()` function and then by applying the `describe()` function on the 
$3, 9, 6, 10, 7, 9, 7, 6, 1$ 
series.

**Note:** The MSD is also called **variance**. Therefore, 
$variance = \sigma^2$
where 
$\sigma$ 
is the standard deviation.

#### Activity 1: Univariate Grouping

Before we begin grouping and aggregation, let's first get the list of columns we are left with after applying all the previous operations.

The description for all the columns containing data for air pollutants, temperature, relative humidity and absolute humidity is provided below.


|Columns|Description|
|-|-|
|PT08.S1(CO)|PT08.S1 (tin oxide) hourly averaged sensor response (nominally $\text{CO}$ targeted)|
|C6H6(GT)|True hourly averaged Benzene concentration in $\frac{\mu g}{m^3}$|
|PT08.S2(NMHC)|PT08.S2 (titania) hourly averaged sensor response (nominally $\text{NMHC}$ targeted)|
|PT08.S3(NOx)|PT08.S3 (tungsten oxide) hourly averaged sensor response (nominally $\text{NO}_x$ targeted)|
|PT08.S4(NO2)|PT08.S4 (tungsten oxide) hourly averaged sensor response (nominally $\text{NO}_2$ targeted)|
|PT08.S5(O3) |PT08.S5 (indium oxide) hourly averaged sensor response (nominally $\text{O}_3$ targeted)|
|T|Temperature in °C|
|RH|Relative Humidity (%)|
|AH|AH Absolute Humidity|

**The `groupby()` Function**

Now, let's group all the values by months, i.e., club all the values together for January, February and so on...

To group the values of a DataFrame by a particular value, use the `groupby()` function. Inside the function, you have to pass the column name to `by` parameter to specify the column about which you want to group the values.

**Syntax:** `dataframe.groupby(by='column')`

You can additionally set the `sort` parameter either equal to `True` or `False` depending on whether you want to sort the values or not. By default, the `groupby()` function returns a new Pandas `DataFrameGroupBy` object containing the grouped sorted values. 

The `group_by()` function returns a `DataFrameGroupBy` object with its physical memory location.

How you want to group the values depends on the problem statement at hand. Here, we would like to analyse data month-wise. Hence, we have grouped the DataFrame by the `Month` column.

**The `get_group()` Function**

To get all the occurrences of a value from a group, use the `get_group()` function. 

**Syntax:** `DataFrameGroupBy_object.get_group(item)`

E.g., if you want to get all the records for the month of March from the `group_2004_month DataFrameGroupBy` object, pass `3` as input to the `get_group()` function.

The above DataFrame contains all the records only for April 2004.

Similarly, you can get values for all other months from the `grouped_2004_month DataFrameGroupBy` object.

Now, you group the records of 2005 about the `Month` column so that we can later compare the situation in 2004 with 2005.

#### Activity 2: Aggregation on Univariate Grouping^

You have already learnt the aggregation functions such as `count(), sum(), min(), max(), mean()` etc. Let's apply a few of them on the `DataFrameGroupBy` objects that we have already created.

The pollution is most severe during the winter season in any part of the world. The winter season in Italy begins from the last week of October and lasts till the last week of March. So, let's get the descriptive statistics for the concentrations of the air pollutants, temperature, relative humidity and absolute humidity in March, November and December in 2004 and 2005.

If we keep on aggregating the values month-wise one-by-one, then it will become a cumbersome task. Let's get the descriptive statistics for all the months for both the years 2004 and 2005.

We still have quite large DataFrames containing the descriptive statistics for all the months and all the columns. Let's try to retrieve only the mean, standard deviation and median values for all the months.

#### Activity 3: The `agg()` Function^^^

Instead of applying the aggregation functions on a `DataFrameGroupBy` object individually, you can apply them together using the `agg()` function. Inside the `agg()` function, you have to pass the names of the functions (as string values) as inputs.

**Syntax:** `DataFrameGroupBy.agg(func=('func1', 'func2' ... 'funcN'))`

where `'func1', 'func2' ... 'funcN'` are the names of the aggregation functions to be applied. They are passed as a singular value (through tuple) to the `func` parameter.

Let's apply the `mean(), std()` and `median()` function on the `group_2004_month DataFrameGroupBy` object.

**Note:** You also choose to apply different aggregation functions on different columns by passing a dictionary containing the names of the columns and the aggregation function(s) to be applied on them.

E.g., let's apply the `mean(), std()` & `median()` functions on the `T` (temperature) column and `max()` function on the `AH` (Absolute humidity) column.

#### Activity 4: Slicing `DataFrameGroupBy` Object^^


We still want to know the mean, standard deviation and median values only for the winter season for 2004 and 2005. 

So, we can use the `loc[]` function to get the rows only for the winter season.

But we don't want any of the aggregated values for the `Year` and `Day` columns. Hence, we need to specify the column indices for the columns we want to be displayed.

Here, we have a collection of column names. For each column, the column itself and the subcolumns (`mean, std, median`) are put together in a tuple. So, for every column, we have three tuples. 

We don't want the last six combinations of columns and sub columns to be part of our final aggregated DataFrame. Hence, we will pass the remaining column indices to the `loc[]` function.

The `group_2004_month.agg(func=['mean', 'std', 'median']).columns[:-6]` is a list of combinations of all the columns and sub columns except for the last 6 combinations.

**Note:** Unfortunately, the `quantile()` function cannot be applied as an aggregation function on a `DataFrameGroupBy` object. To get quartiles, use the `describe()` function.

Let's have a look on how we can get only quartiles using the `describe()` function.

First, we have applied the `describe()` function which will return all the descriptive statistics for the `DataFrameGroupBy` object. Then using the `loc[]` function we have retrieved only the quartile values for the winter season of 2004.

**Note:** Since we have grouped the DataFrames for the years 2004 and 2005 by the `Month` column, it becomes the default index column in the aggregated DataFrames. If you don't want it to happen, pass the `as_index=False` parameter inside the `groupby()` function. 

#### Activity 5: Multivariate Grouping & Aggregation

We can also group multiple columns at once by passing a list of columns to be grouped.

Let's group the DataFrames about the `Month` & `Day Name` columns.

Let's get all the records for all Mondays in the month of March, 2004. 

**Note:** The values to be retrieved should be passed as a tuple in the same order of grouping.

**Python Pandas - Merging/Joining**

Pandas has full-featured, high performance in-memory join operations idiomatically very similar to relational databases like SQL.

Pandas provides a single function, merge, as the entry point for all standard database join operations between DataFrame objects −

**Merge Two DataFrames on a Key**


**Merge Two DataFrames on Multiple Keys**

**Python Pandas - Concatenation**

Concatenating Objects
The concat function does all of the heavy lifting of performing concatenation operations along an axis. Let us create different objects and do concatenation.

Suppose we wanted to associate specific keys with each of the pieces of the chopped up DataFrame. We can do this by using the keys argument −

The index of the resultant is duplicated; each index is repeated.

If the resultant object has to follow its own indexing, set ignore_index to True.

Observe, the index changes completely and the Keys are also overridden.

If two objects need to be added along axis=1, then the new columns will be appended.